### PR TITLE
Add more types to property values reported to UI

### PR DESCRIPTION
**Pull request by hunger**

### CODE REVIEW

File: tools/lsp/preview/ui.rs [hunk 1]
```diff
@@ -253,39 +258,156 @@ fn extract_tr_data(tr_node: &syntax_nodes::AtTr) -> Option<(String, String, Stri
...
+            extract_value_with_unit(&expression, &[Unit::S, Unit::Ms], &mut value)
+        }
+        langtype::Type::PhysicalLength => {
+            extract_value_with_unit(&expression, &[Unit::Phx], &mut value)
+        }
+        langtype::Type::LogicalLength => extract_value_with_unit(
+            &expression,
+            &[Unit::Px, Unit::Cm, Unit::Mm, Unit::In, Unit::Pt],
+            &mut value,
+        ),
```
[the remark this answers is not in the document]
Comment (hunger) [reply tools/lsp/preview/ui.rs:357]: I was wondering whether that would make sense, but just stuck to the basics for a start. Ill change that!

File: tools/lsp/preview/ui.rs [hunk 2]
```diff
@@ -253,39 +258,156 @@ fn extract_tr_data(tr_node: &syntax_nodes::AtTr) -> Option<(String, String, Stri
...
+                    if let Some(color) = literals::parse_color_literal(&text) {
+                        value.kind = PropertyValueKind::Color;
+                        value.value_alpha = ((color & 0xff000000) >> 24) as i32;
+                        value.value_red = ((color & 0x00ff0000) >> 16) as i32;
+                        value.value_green = ((color & 0x0000ff00) >> 8) as i32;
+                        value.value_blue = (color & 0x000000ff) as i32;
+                    } else {
+                        // TODO: Extract `Foo.bar` as Palette `Foo`, entry `bar`.
+                        // This makes no sense right now, as we have no way to get any
+                        // information on the palettes.
```
[the remark this answers is not in the document]
Comment (hunger) [reply tools/lsp/preview/ui.rs:393]: Without information on which palettes are there and which colors they contain, the information that some property is using a color `bar` from palette `foo` is not too helpful.

Yes, I could add the actual color value, but we should not replace `Colors.red` with `#fff8000000` -- at least not without offering a way to pick another color from a palette.

File: tools/lsp/preview/ui.rs [hunk 3]
```diff
@@ -253,39 +258,156 @@ fn extract_tr_data(tr_node: &syntax_nodes::AtTr) -> Option<(String, String, Stri
...
+                        value.kind = PropertyValueKind::Integer;
+                        value.value_int = v as i32;
+                    }
+                }
+            } else if value.code.is_empty() {
+                value.kind = PropertyValueKind::Integer;
+            }
+        }
+        langtype::Type::Color => {
+            eprintln!("Looking at color!");
```
[the remark this answers is not in the document]
Comment (hunger) [reply tools/lsp/preview/ui.rs:381]: We will have a simpler UI for a `Color` than a `Brush`. So we need to tell the UI whether we have one or the other. So I thought I'd keep them separate for now.

How would you transfer the ARGB information across? Just stick the u32 into an int and be done with that?

[the remark this answers is not in the document]
Comment (ogoffart) [reply tools/lsp/preview/ui.rs:381]: Ok, the color and brush should have different UI (no gradient option for color)

> How would you transfer the ARGB information across

via the `color.red`, `color.alpha`, ...  (https://releases.slint.dev/1.7.2/docs/slint/src/language/syntax/types.html#properties)

[the remark this answers is not in the document]
Comment (hunger) [reply tools/lsp/preview/ui.rs:381]: D*oh! Of course, I can pass a Color! I was so focused on simple types. I have worked way too long within the LSP protocol;-)

File: tools/lsp/preview/ui.rs (hunk 1 above)
[the remark this answers is not in the document]
Comment (hunger) [reply tools/lsp/preview/ui.rs:357]: Will be in the next push.

File: tools/lsp/ui/api.slint [hunk 4]
```diff
@@ -70,25 +70,32 @@ export struct Range {
...
 export enum PropertyValueKind {
-    Code,
     Boolean,
-    String,
+    Code,
+    Color,
     Enum,
+    Float,
+    Integer,
+    String,
```
[the remark this answers is not in the document]
Comment (hunger) [reply tools/lsp/ui/api.slint:79]: The next update will have these renamed.

File: tools/lsp/preview/ui.rs (hunk 1 above)
[the remark this answers is not in the document]
Comment (hunger) [reply tools/lsp/preview/ui.rs:357]: The next update will have this.

**Comment by hunger** [comment]
I added Brush as a type, but so far I am only handling the solid color option there.

Gradients will need a bit of twiddling to parse correctly. The code doing that seems to life in the resolving step -- and resolves expressions. I need to not do that.

**Pull request merged**